flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/collect/CollectSinkOperatorCoordinator.java

```diff
@@ -0,0 +1,212 @@
+import org.slf4j.LoggerFactory;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.net.InetSocketAddress;
+import java.net.Socket;
+import java.util.Collections;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+/**
+ * {@link OperatorCoordinator} for {@link CollectSinkFunction}.
+ *
+ * <p>This coordinator only forwards requests and responses from clients and sinks,
+ * it does not store any state in itself.
+ */
+public class CollectSinkOperatorCoordinator implements OperatorCoordinator, CoordinationRequestHandler {
+
+	private static final Logger LOG = LoggerFactory.getLogger(CollectSinkOperatorCoordinator.class);
+
+	private InetSocketAddress address;
+	private Socket socket;
+
+	private ExecutorService executorService;
+
+	@Override
+	public void start() throws Exception {
+		this.executorService = Executors.newFixedThreadPool(1);
+	}
+
+	@Override
+	public void close() throws Exception {
+		closeCurrentConnection();
+		this.executorService.shutdown();
+	}
+
+	@Override
+	public void handleEventFromOperator(int subtask, OperatorEvent event) throws Exception {
+		Preconditions.checkArgument(
+			event instanceof CollectSinkAddressEvent, "Operator event must be a CollectSinkAddressEvent");
+		address = ((CollectSinkAddressEvent) event).getAddress();
+		LOG.debug("Received sink socket server address: " + address);
+
+		// this event is sent when the sink function starts, so we remove the old socket if it is present
+		socket = null;
+	}
+
+	@Override
+	public CompletableFuture<CoordinationResponse> handleCoordinationRequest(CoordinationRequest request) {
+		Preconditions.checkArgument(
+			request instanceof CollectCoordinationRequest,
+			"Coordination request must be a CollectCoordinationRequest");
+
+		CompletableFuture<CoordinationResponse> responseFuture = new CompletableFuture<>();
+		executorService.submit(() -> handleRequestImpl((CollectCoordinationRequest) request, responseFuture));
+		return responseFuture;
+	}
+
+	private void handleRequestImpl(
+			CollectCoordinationRequest request,
+			CompletableFuture<CoordinationResponse> responseFuture) {
+		// there is no thread safety issue in this method
+		// because the thread pool has only 1 thread
+		// and by design there shall be only 1 client reading from this coordinator
+		if (address == null) {
+			responseFuture.complete(new CollectCoordinationResponse(createEmptySerializedResult(request)));
+		}
+
+		try {
+			if (socket == null) {
+				socket = new Socket(address.getAddress(), address.getPort());
+				LOG.debug("Sink connection established");
+			}
+
+			// send version and token to sink server
+			LOG.debug("Forwarding request to sink socket server");
+			DataOutputViewStreamWrapper outStream = new DataOutputViewStreamWrapper(socket.getOutputStream());
```

Comment:
after socket established, this `outStream` and `inStream` could be reused and don't have to create new instance each time